File: lib/responseParser.ts
import type { RoundtablePersonaResponse, RoundtableResponse } from "./types";

const validPersonaIds = new Set([
  "value_allocator",
  "market_structure_trader",
  "behavioral_contrarian",
]);

export type ParsedRoundtable =
  | {
      ok: true;
      data: RoundtableResponse;
    }
  | {
      ok: false;
      error: string;
    };

function isPersonaResponse(value: unknown): value is RoundtablePersonaResponse {
  if (!value || typeof value !== "object") {
    return false;
  }
  const item = value as Record<string, unknown>;
  return (
    typeof item.persona_id === "string" &&
    validPersonaIds.has(item.persona_id) &&
    typeof item.persona_name === "string" &&
    typeof item.message === "string" &&
    item.message.trim().length > 0
  );
}

export function parseRoundtableJson(raw: string): ParsedRoundtable {
  try {
    const parsed = JSON.parse(raw) as unknown;
    if (!parsed || typeof parsed !== "object") {
      return { ok: false, error: "Roundtable response was not a JSON object." };
    }
    const record = parsed as Record<string, unknown>;
    if (!Array.isArray(record.responses)) {
      return { ok: false, error: "Roundtable response did not include a responses array." };
    }
    const responses = record.responses.filter(isPersonaResponse);
    if (responses.length === 0) {
      return { ok: false, error: "Roundtable response did not include valid persona messages." };
    }
    return {
      ok: true,
      data: {
        responses,
        learning_takeaway:
          typeof record.learning_takeaway === "string" ? record.learning_takeaway : "",
      },
    };
  } catch {
    return { ok: false, error: "Roundtable response was not valid JSON." };
  }
}

File: src/response_parser.py
import json
import re

VALID_PERSONA_IDS = {
    "value_allocator",
    "market_structure_trader",
    "behavioral_contrarian",
}


def parse_roundtable_json(raw_response: str) -> dict:
    """Parse roundtable JSON and return normalized persona messages."""
    try:
        data = json.loads(raw_response)
    except json.JSONDecodeError as exc:
        return {
            "responses": [],
            "learning_takeaway": "",
            "parse_error": f"Invalid JSON: {exc}",
            "raw_response": raw_response,
        }

    if not isinstance(data, dict):
        return {
            "responses": [],
            "learning_takeaway": "",
            "parse_error": "Roundtable response must be a JSON object.",
            "raw_response": raw_response,
        }

    raw_responses = data.get("responses", [])
    if not isinstance(raw_responses, list):
        return {
            "responses": [],
            "learning_takeaway": "",
            "parse_error": "Roundtable response field 'responses' must be a list.",
            "raw_response": raw_response,
        }

    normalized = []
    for item in raw_responses:
        if not isinstance(item, dict):
            continue
        persona_id = str(item.get("persona_id", "")).strip()
        persona_name = str(item.get("persona_name", "")).strip()
        message = str(item.get("message", "")).strip()
        if persona_id in VALID_PERSONA_IDS and persona_name and message:
            normalized.append(
                {
                    "persona_id": persona_id,
                    "persona_name": persona_name,
                    "message": message,
                }
            )

    if not normalized:
        return {
            "responses": [],
            "learning_takeaway": "",
            "parse_error": "No valid persona responses found.",
            "raw_response": raw_response,
        }

    return {
        "responses": normalized,
        "learning_takeaway": str(data.get("learning_takeaway", "")).strip(),
        "parse_error": "",
        "raw_response": raw_response,
    }


def split_roundtable_response(response: str, persona_names: list[str]) -> dict[str, str]:
    """Split a markdown roundtable response into persona sections."""
    sections: dict[str, str] = {}
    headings = persona_names + ["Final Learning Takeaway"]
    escaped = "|".join(re.escape(name) for name in headings)
    pattern = re.compile(rf"^##\s+({escaped})\s*$", flags=re.MULTILINE)
    matches = list(pattern.finditer(response))

    for index, match in enumerate(matches):
        title = match.group(1)
        start = match.end()
        end = matches[index + 1].start() if index + 1 < len(matches) else len(response)
        content = response[start:end].strip()
        if content:
            sections[title] = content

    if not sections:
        sections["Final Learning Takeaway"] = response.strip()

    return sections
